p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java

```diff
@@ -0,0 +1,190 @@
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Has business logic to operate data stores which are spread across multiple files.<br><br><p>
+ *
+ * <h3>Use Case</h3>
+ * Startup requires to send all object keys to the network in order to get the new data.
+ * However, data stores got quite big over time and will grow even faster in the future.
+ * With multi-file data stores, we can query for new objects since the last snapshot (ie.
+ * the last release) and subsequently send "I am Bisq vx.y.z and I have these new objects"
+ * to the network and the network can respond accordingly.</p><br><br><p>
+ *
+ * <h3>Features</h3>
+ * In order to only get a specific part of all the objects available in the complete data
+ * store, the original single-file data store had to be split up and this class and
+ * {@link SplitStore} are there to handle the business logic needed for
+ * <ul>
+ * <li>migrating to the new and shiny multi-file data store</li>
+ * <li>shoveling around data in case Bisq gets updated to a new version</li>
+ * <li>takes care of setting up a fresh Bisq install</li>
+ * <li>makes sure that historical data cannot be altered easily</li>
+ * <li>adding data to the store will only amend the live store</li>
+ * <li>takes care of keeping the legacy API functional</li>
+ * <li>adds the feature of filtering object queries by Bisq release</li>
+ * </ul>
+ * </p>
+ *
+ * <h3>Further reading</h3><p><ul>
+ *     <li><a href="https://github.com/bisq-network/projects/issues/25">project description</a></li>
+ * </ul></p>
+ */
+@Slf4j
+public abstract class SplitStoreService<T extends SplitStore> extends MapStoreService<T, PersistableNetworkPayload> {
+    protected HashMap<String, SplitStore> history;
+
+    public SplitStoreService(File storageDir, Storage<T> storage) {
+        super(storageDir, storage);
+    }
+
+    @Override
+    protected void put(P2PDataStorage.ByteArray hash, PersistableNetworkPayload payload) {
+        // make sure we do not add data that we already have (in a bin of historical data)
+        if (getMap().containsKey(hash))
+            return;
+
+        store.getMap().put(hash, payload);
+        persist();
+    }
+
+    @Override
+    protected PersistableNetworkPayload putIfAbsent(P2PDataStorage.ByteArray hash,
+                                                    PersistableNetworkPayload payload) {
+        // make sure we do not add data that we already have (in a bin of historical data)
+        if (getMap().containsKey(hash))
+            return null;
+
+        PersistableNetworkPayload previous = store.getMap().putIfAbsent(hash, payload);
+        persist();
+        return previous;
+    }
+
+
+    @Override
+    public Map<P2PDataStorage.ByteArray, PersistableNetworkPayload> getMap() {
+        HashMap<P2PDataStorage.ByteArray, PersistableNetworkPayload> result = new HashMap<>();
+        result.putAll(store.getMap());
+        history.forEach((s, store) -> result.putAll(store.getMap()));
```

Comment:
I thought about that. The only thing that can be cached is loading the historical data stores into the result. And since these are read-only anyways, there is nothing that changes. So why not cache it?

Well, the maps of the stores are in memory as well. So the data is already there and ready to be transferred. The only redirection that happens when we do it like this is that we iterate through all stores and add the data to the result. Compare that to a "cached" map of historical data, the data has to be transferred as well. The only thing we gain by caching the data is we need more RAM while saving on a few instructions to iterate through history.

IMO The small computational overhead does outweigh adding the complexity of another caching layer that needs to be maintained.

Anyways, this piece of code is going to be deprecated by https://github.com/bisq-network/projects/issues/29 anyways. Then is the time IMO to really dial down performance.